### PR TITLE
Make changes to models to store IDs

### PR DESCRIPTION
**Pull request by vishneshjr**
# Refactored Models and Firebase

## Description

- Models are refactored so that they do not reference other objects or lists of objects. They only reference object IDs or lists of object IDs. Models can optionally store the actual object or list of objects. These can be fetched from Firebase when required. Also made all the models structs.

- Added `@DocumentID` tag to id of models so that their firebase id is the same as their id parameter.

- Added coding keys so that all models can directly be encoded or decoded without having to use dictionaries. Removed all manual dictionary creation from the FirebaseDataManager

- Change sample object instantiation so that they always have objects when created.

### CODE REVIEW

**Comment by JosheevR** [comment]
Firebase adding for groups and object data should still work. Fetching shouldn't work anymore with your changes, and should be fixed.

**Comment by JosheevR** [comment]
Add score int to Group model.

**Comment by JosheevR** [comment]
Save group id for as a field in each group in Firebase.

File: Vineyard/Vineyard/Badge.swift
Comment (rahulon12) [inline Vineyard/Vineyard/Badge.swift:20]: Did you double check whether explicitly specifying the `CodingKeys` means the other properties are not encoded/decoded?

File: Vineyard/Vineyard/GroupsListViewModel.swift
Comment (rahulon12) [inline Vineyard/Vineyard/GroupsListViewModel.swift]: Did you see whether the group updates live upon changes?

File: Vineyard/Vineyard/Person.swift
Comment (rahulon12) [inline Vineyard/Vineyard/Person.swift:33]: I would use a `guard` here and make sure the `id` is present instead of generating a new `id`.

```swift
mutating func addGroup(_ group: Group) {
    guard let id = group.id else { fatalError(/* ... */) }

    self.groupIDs.append(id)
}
```

File: Vineyard/Vineyard/Badge.swift
Comment (rahulon12) [inline Vineyard/Vineyard/Badge.swift]: I don't think we need an initial value for anything marked with `@DocumentID`. I'm seeing this warning in the console:
```
The document ID is managed by Firestore and any initialized or set value will be ignored. 
The ID is automatically set when reading from Firestore.
```

```suggestion
    @DocumentID var id: String?
```

**Review by rahulon12 — APPROVED** [review]
I think this is good, a couple of small nits.

File: Vineyard/Vineyard/GroupsListViewModel.swift
Comment (rahulon12) [reply Vineyard/Vineyard/GroupsListViewModel.swift]: If this is not needed now, we can leave this for another PR.

**Comment by rahulon12** [comment]
Wait for #66 before merging

**Comment by rahulon12** [comment]
Ok we're good to go now, once you fix the stuff above

**Comment by sankaet8** [comment]
I will be reviewing once you make the changes that rahul asked, and then check functionality, I think the code looks good tho

**Pull request merged**
